lib/libc/gen/lockf.3
.\" $NetBSD: lockf.3,v 1.11 2011/03/30 16:37:09 jruoho Exp $
.\"
.\" Copyright (c) 1997 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Klaus Klein and S.P. Zeidler.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd March 30, 2011
.Dt LOCKF 3
.Os
.Sh NAME
.Nm lockf
.Nd record locking on files
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft int
.Fn lockf "int filedes" "int function" "off_t size"
.Sh DESCRIPTION
The
.Fn lockf
function allows sections of a file to be locked with advisory-mode locks.
Calls to
.Fn lockf
from other processes which attempt to lock the locked file section will
either return an error value or block until the section becomes unlocked.
All the locks for a process are removed when the process terminates.
.Pp
The argument
.Fa filedes
is an open file descriptor.
The file descriptor must have been opened either for write-only
.Dv ( O_WRONLY )
or read/write
.Dv ( O_RDWR )
operation.
.Pp
The
.Fa function
argument is a control value which specifies the action to be taken.
The permissible values for
.Fa function
are as follows:
.Pp
.Bl -tag -width F_ULOCKXX -compact -offset indent
.It Sy Function
.Sy Description
.It Dv F_ULOCK
unlock locked sections
.It Dv F_LOCK
lock a section for exclusive use
.It Dv F_TLOCK
test and lock a section for exclusive use
.It Dv F_TEST
test a section for locks by other processes
.El
.Pp
.Dv F_ULOCK
removes locks from a section of the file;
.Dv F_LOCK
and
.Dv F_TLOCK
both lock a section of a file if the section is available;
.Dv F_TEST
detects if a lock by another process is present on the specified section.
.Pp
The
.Fa size
argument is the number of contiguous bytes to be locked or
unlocked.
The section to be locked or unlocked starts at the current
offset in the file and extends forward for a positive size or backward
for a negative size (the preceding bytes up to but not including the
current offset).
However, it is not permitted to lock a section that
starts or extends before the beginning of the file.
If
.Fa size
is 0, the section from the current offset through the largest possible
file offset is locked (that is, from the current offset through the
present or any future end-of-file).
.Pp
The sections locked with
.Dv F_LOCK
or
.Dv F_TLOCK
may, in whole or in part, contain or be contained by a previously
locked section for the same process.
When this occurs, or if adjacent
locked sections would occur, the sections are combined into a single
locked section.
If the request would cause the number of locks to
exceed a system-imposed limit, the request will fail.
.Pp
.Dv F_LOCK
and
.Dv F_TLOCK
requests differ only by the action taken if the section is not
available.
.Dv F_LOCK
blocks the calling process until the section is available.
.Dv F_TLOCK
makes the function fail if the section is already locked by another
process.
.Pp
File locks are released on first close by the locking process of any
file descriptor for the file.
.Pp
.Dv F_ULOCK
requests release (wholly or in part) one or more locked sections
controlled by the process.
Locked sections will be unlocked starting
at the current file offset through
.Fa size
bytes or to the end of file if size is 0.
When all of a locked section
is not released (that is, when the beginning or end of the area to be
unlocked falls within a locked section), the remaining portions of
that section are still locked by the process.
Releasing the center
portion of a locked section will cause the remaining locked beginning
and end portions to become two separate locked sections.
If the
request would cause the number of locks in the system to exceed a
system-imposed limit, the request will fail.
.Pp
An
.Dv F_ULOCK
request in which size is non-zero and the offset of the last byte of
the requested section is the maximum value for an object of type
off_t, when the process has an existing lock in which size is 0 and
which includes the last byte of the requested section, will be treated
as a request to unlock from the start of the requested section with a
size equal to 0.
Otherwise an
.Dv F_ULOCK
request will attempt to unlock only the requested section.
.Pp
A potential for deadlock occurs if a process controlling a locked
region is put to sleep by attempting to lock the locked region of
another process.
This implementation detects that sleeping until a
locked region is unlocked would cause a deadlock and fails with an
.Er EDEADLK
error.
.Pp
.Fn lockf ,
.Xr fcntl 2
and
.Xr flock 2
locks may be safely used concurrently.
.Pp
Blocking on a section is interrupted by any signal.
.Sh RETURN VALUES
If successful, the
.Fn lockf
function returns 0.
Otherwise, it returns \-1, sets
.Va errno
to indicate an error, and existing locks are not changed.
.Sh ERRORS
.Fn lockf
will fail if:
.Bl -tag -width Er
.It Bq Er EAGAIN
The argument
.Fa function
is
.Dv F_TLOCK
or
.Dv F_TEST
and the section is already locked by another process.
.It Bq Er EBADF
The argument
.Fa filedes
is not a valid open file descriptor.
.Pp
The argument
.Fa function
is
.Dv F_LOCK
or
.Dv F_TLOCK ,
and
.Fa filedes
is not a valid file descriptor open for writing.
.It Bq Er EDEADLK
The argument
.Fa function
is
.Dv F_LOCK
and a deadlock is detected.
.It Bq Er EINTR
The argument
.Fa function
is F_LOCK
and
.Fn lockf
was interrupted by the delivery of a signal.
.It Bq Er EINVAL
The argument
.Fa function
is not one of
.Dv F_ULOCK ,
.Dv F_LOCK ,
.Dv F_TLOCK
or
.Dv F_TEST .
.Pp
The argument
.Fa filedes
refers to a file that does not support locking.
.It Bq Er ENOLCK
The argument
.Fa function
is
.Dv F_ULOCK ,
.Dv F_LOCK
or
.Dv F_TLOCK ,
and satisfying the lock or unlock request would result in the number
of locked regions in the system exceeding a system-imposed limit.
.El
.Sh SEE ALSO
.Xr fcntl 2 ,
.Xr flock 2
.Sh STANDARDS
The
.Fn lockf
function conforms to
.St -xpg4.2 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.1537 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 5.0 to NetBSD 6.0:

	syslog(3): Import Martin Schuette's Summer of Code project:
		Reliable TCP connections, syslog protocol API,
		encryption. [christos 20081031]
	agp(4): Add support for Intel G35. [christos 20081108]
	kernel: Add generic producer/consumer queue. [matt 20081111]
	lkm: Removed. Superseded by the new module(7) framework. [ad 20081112]
	module(7): Allow the POSIX semaphore code to be loaded as a module.
		[ad 20081112]
	module(7): Make compat_ossaudio a module. [ad 20081113]
	module(7): Make compat_ndis a module. [ad 20081113]
	build.sh: Add support for new MAKEVERBOSE levels 3 and 4.
		[apb 20081113]
	libfetch: Update to libfetch 2.20. Allow HTTP auth to be provided in
		URL. [joerg 20081114]
	module(7): Make POSIX AIO loadable as a module. [ad 20081114]
	amd64, i386: Make boot loader autoload a kernel module corresponding
		to the root file system type. [ad 20081119]
	module(7): Make the emulations, exec formats, coredump, NFS, and
		the NFS server into modules. [ad 20081119]
	rump(3): Move VFS functionality away from librump to a new library:
		librumpvfs [pooka 20081119]
	kernel, libc: Add new ptree (Patricia / RADIX tree) implementation.
		[matt 20081120]
	i386: Switch the i386 installation CDs away from RAM disks and
		instead use the CD-ROM as the root file system. [ad 20081121]
	i386: Remove most compat and file system options from GENERIC
		since they are also provided by modules. [ad 20081122]
	nfs: Avoid noisy nfs_timer/nfs_reply DEBUG output that occurs
		when the NFS server goes away. [mrg 20081123]
	compat_osf1(8): Add getpgid, needed for Tru64 4.0g ksh.
		[njoly 20081124]
	module(7): Make PPP compressors modules. [cube 20081125]
	i386: Make pxeboot(8) understand boot.cfg(5). [ad 20081125]
	agp(4): Add support for the Intel G45 AGP. [christos 20081129]
	compat_osf1(8): Fix getdirentries(2). [njoly 20081202]
	module(7): Add a couple of sysctl(8) entries, kern.module.autoload and
		kern.module.verbose. [ad 20081203]
	module(7): Add a layerfs module. [ad 20081205]
	libc: Add fdopendir(3), from Solaris/Linux. [ad 20081205]
	compat_linux32: Add getresuid() and getresgid() syscalls.
		[njoly 20081205]
	ddb(4): Add 'show vmem' and 'show all vmems' commands.
		[cegger 20081207]
	re(4): Add support for a 8168C/8111C revision. [alc 20081208]
	tab(1): Add tabs utility, fixes PR bin/11227. [roy 20081211]
	httpd(8): Add rc.d script to control httpd daemon. [mishka 20081211]
	ath(4): Switch from the binary HAL to the open-source HAL released by
		Sam Leffler. [alc 20081211]
	compat_linux32: Add clock_nanosleep syscall. [njoly 20081212]
	boot.cfg(5): Add an option to clear the screen. [christos 20081213]
	pms(4): Add Elantech multi-touch trackpad support [jmcneill 20081213]
	sparc: Switch to use wscons. [macallan 20081213]
	snprintb(3), kprintf(9): Share snprintb(3) among kernel and libutil
		and make it supersede bitmask_snprintf(9). [christos 20081216]
	sup(1): Add -M user flag to specify the username to send mail to from
		the command line. [christos 20081217]
	malloc(9): Remove MALLOC() and FREE() macros. [cegger 20081217]
	x68k:	pow(4) switches to use MI powerd(8) [isaki 20081220]
	viaide(4): Add VT8237S Integrated SATA Controller support.
		[nonaka 20081221]
	lvm(8):	Add LVM functionality to NetBSD; this is done by dm(4) driver
		in kernel and lvm2 tools made by RedHat. We use the same well
		known interface as Linux does for managing LVM. [haad 20081222]
	gcscaudio(4): Add AMD Geode CS5536 audio driver, from SHIMIZU Ryo.
		[jmcneill 20081228]
	rump: i386 and amd64 can use kernel module binaries [pooka 20081230]
	mount_cd9660(8): Add nocasetrans option as a synonym for nomaplcase.
		[christos 20090103]
	rump: Include libkern in rumpkern. [pooka 20090104]
	rc.d(8): Add a script for rndctl(8). [apb 20090104]
	module(7): Add sysvbfs module. [pooka 20090105]
	rump: Prepend all symbols within the rump virtual kernel image
		with the string "rump".  This prevents conflicts with the
		host system symbols.  [pooka 20090108]
	bwi(4): Add a driver for Broadcom AirForce / Apple Airport Extreme
		wireless network controllers. From Taylor Campbell.
		[macallan 20090109]
	rgephy(4): Support for RTL8211C(L) phy. From FreeBSD. [cegger 20090109]
	rgephy(4): Support for flow control. From OpenBSD. [cegger 20090109]
	makefs(8): Add support for encoding RISC OS metadata into
		ISO 9660 file systems. [bjh21 20090110]
	time_t, dev_t: Change time_t and dev_t to be 64 bit quantities. Fixed
		POSIX issues in timeval and timespec. [christos 20090110]
	make(1): Changes to the way .for loops are processed. [dsl 20090111]
	mount(8): Add -o rump, which mounts the file system as a userspace
		server via puffs and rump instead of using a
		kernel driver. [pooka 20090111]
	openpgpsdk: Initial import of openpgpsdk v0.9. [agc 20090112]
	rump: Implement a virtual block device driver.  [pooka 20090113]
	rump: Support __RENAME in syscall interfaces and expose the

regress/sys/kern/lockf/lockf.c
/*	$NetBSD: lockf.c,v 1.6.18.1 2008/05/18 12:30:49 yamt Exp $	*/

/*-
 * Copyright (c) 2000 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * lockf regression test:
 *
 * Tests:
 * 1) fork N child processes, do a bunch of random byte range lock/unlock.
 */

#include <sys/types.h>
#include <sys/wait.h>
#include <sys/ptrace.h> 

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <err.h>
#include <signal.h>
#include <errno.h>

int nlocks = 1000;		/* number of locks per thread */
int nprocs = 10;		/* number of processes to spawn */
int sleeptime = 500000;		/* sleep time between locks, usec */
off_t size = 8192;		/* size of file to lock */
const char *lockfile = "/tmp/lockf_test";

static u_int32_t
random_uint32(void)
{
	return lrand48();
}


static void
trylocks(int id)
{
	int i, ret, fd;
	const char *which = NULL;
	
	srand48(getpid());

	fd = open (lockfile, O_RDWR, 0);
	
	if (fd < 0)
		err(1, lockfile);
	
	printf("%d: start\n", id);
	
	for (i=0; i<nlocks; i++) {
		struct flock fl;

		fl.l_start = random_uint32() % size;
		fl.l_len = random_uint32() % size;
		switch (random_uint32() % 3) {
		case 0:
			which = "read";
			fl.l_type = F_RDLCK;
			break;
		case 1:
			which = "write";
			fl.l_type = F_WRLCK;
			break;
		case 2:
			which = "un";
			fl.l_type = F_UNLCK;
			break;
		}
		fl.l_whence = SEEK_SET;

		printf("%d: try %slock %d to %d\n", id, which, (int)fl.l_start,
		    (int)(fl.l_start + fl.l_len));
		
		ret = fcntl(fd, F_SETLKW, &fl);

		if (ret < 0)
			perror("fcntl");
		printf("%d: got %slock %d to %d\n", id, which, (int)fl.l_start,
		    ((int)(fl.l_start + fl.l_len)));
		
		if (usleep(sleeptime) < 0) 
		  err(1, "usleep");
	}
	printf("%d: done\n", id);
	close (fd);
}

/* ARGSUSED */
int
main(int argc, char **argv)
{
	int i, j;
	pid_t *pid;
	int status;
	int fd;
	
	(void)unlink(lockfile);

	fd = open (lockfile, O_RDWR|O_CREAT|O_EXCL|O_TRUNC, 0666);
	if (fd < 0)
		err(1, "%s", lockfile);

	if (ftruncate(fd, size) < 0)
		err(1, "ftruncate of %s failed", lockfile);

	fsync(fd);
	close(fd);
	
	pid = malloc(nprocs * sizeof(pid_t));
	
	for (i=0; i<nprocs; i++) {
		pid[i] = fork();
		switch (pid[i]) {
		case 0:
			trylocks(i);
			_exit(0);
			break;
		case -1:
			err(1, "fork failed");
			break;
		default:
			break;
		}
	}
	for (j=0; j<100; j++) {
		printf("parent: run %i\n", j+1);
		for (i=0; i<nprocs; i++) {
			printf("stop %d\n", i);
			if (ptrace(PT_ATTACH, pid[i], 0, 0) < 0)
				err(1, "ptrace attach %d", pid[i]);
			printf("wait %d\n", i);
			if (waitpid(pid[i], &status, WUNTRACED) < 0)
				err(1, "waitpid(ptrace)");
			printf("awake %d\n", i);
			usleep(sleeptime/3);
			if (ptrace(PT_DETACH, pid[i], (caddr_t)1, 0) < 0)
				err(1, "ptrace detach %d", pid[i]);
			printf("done %d\n", i);
			usleep(sleeptime/3);
		}
	}
	for (i=0; i<nprocs; i++) {
		printf("reap %d: ", i);
		fflush(stdout);
		kill(pid[i], SIGINT);
		waitpid(pid[i], &status, 0);
		printf(" status %d\n", status);
	}
	exit(0);
	/* NOTREACHED */
}
